### ISSUE TITLE
[css-box-3][css-21] Positioning boxes with { float:left; width:0px; }

### ISSUE THREAD

**Issue by FremyCompany** [w3c/csswg-drafts#576]
A float is positioned according to various rules, whose following are relevant in this case: 
> • The outer top of an element's floating box may not be higher than the top of any line-box 
containing a box generated by an element earlier in the source document. 
> • A floating box must be placed as high as possible. 
> • A left-floating box must be put as far to the left as possible, a right-floating box as far to the right 
as possible. A higher position is preferred over one that is further to the left/right. 

I guess how most browsers implement the first rule in code is: 
> Whenever the line builder has to position a float whose width is {NECESSARY}px in a line box that is 
already {COMMITTED}px wide, in a constrained space of {AVAILABLE}px, then the float can be placed to 
the left of the line box if NECESSARY ≤ AVAILABLE - COMMITTED. 

This isn't entirely true, though. In the case of 0px-wide floaters, it is ALWAYS possible to put the floater on the same line as the current one. This is what Edge implements. This is not what most others seems to be doing in the case the line is "already too big" (for instance contains an unbreakable word or a picture that is longer than the container width). 

In short, most browsers just decide to push the float beneath the current line box based on the fact that 
``` 0 ≤ 200 - 300 ⇔ 0 ≤ -100``` (a very well-known corollary of the Zero Theorem). This is not a correct implementation of CSS 2.1 rules. 

Should we file bugs on browsers here, or should we change the rules to match Chrome and Firefox?

**Comment by FremyCompany** [comment]
Here are test cases to play with the behavior: 

-  https://jsfiddle.net/swntbxpd/ 
(remaining space on the line = 0px, so the 0px float can fit on that same line)

-  https://jsfiddle.net/swntbxpd/1/ 
(remaining space on the line = -100px, so the 0px float cannot fit on that same line)

**Comment by dbaron** [comment]
I haven't thought through the whole thing here, but one initial comment:

> In the case of 0px-wide floaters, it is ALWAYS possible to put the floater on the same line as the current one.

This isn't true, because the anchor point of the float need not be at a break point, and the position of the next break point might not fit in the line.  In this case, the float needs to be pushed down in order to satisfy rule 6:
> The outer top of an element's floating box may not be higher than the top of any line-box containing a box generated by an element earlier in the source document. 

It's not yet clear to me how much of the above argument this refutes.

**Comment by dbaron** [comment]
(Although if you're looking at positioning floats being an operation that doesn't happen until you've reached the next break opportunity after their anchor point (placeholder), then it is true.  But it didn't sound like that was the approach you were describing.)

**Comment by dbaron** [comment]
OK, after looking at the second testcase a bit more (which as far as I can tell is the one that differs), I think what this really comes down to is the meaning of [this wording in section 9.5](https://drafts.csswg.org/css2/visuren.html#floats):

>A line box is next to a float when there exists a vertical position that satisfies all of these four conditions: (a) at or below the top of the line box, (b) at or above the bottom of the line box, (c) below the top margin edge of the float, and (d) above the bottom margin edge of the float.
>
>Note: this means that floats with zero outer height or negative outer height do not shorten line boxes.
>
>If a shortened line box is too small to contain any content, then the line box is shifted downward (and its width recomputed) until either some content fits or there are no more floats present.

In particular, does a 0-width float shorten a line box by 0, or does it not shorten the line box.

(It's already stated that a 0-height float does not shorten a line box -- something I suspect Gecko may not have implemented -- or may not have implemented the implications of for float placement rather than line placement.)

**Comment by antonprowse** [comment]
We've had a similar discussion in the past:
https://lists.w3.org/Archives/Public/www-style/2010Aug/0182.html
https://lists.w3.org/Archives/Public/www-style/2010Sep/0129.html

In that discussion we concluded that a 0-width float doesn't shorten a line box.  (Really, we should be a bit more careful here, as we were in that discussion and other contemporary ones; the relevant thing is not the width per se but rather the position of the right margin edge of the float [assuming it's a left float] and whether that intrudes a non-zero amount into the space where the line box would otherwise begin.  The final two examples in the first post I linked to illustrate that, and also illustrate that browsers don't regard the float as shortening the line box [else the subsequent overflowing unbreakable word would have to be shunted down to below the float].)

But let's quickly consider what would happed if such a float _is_ regarded as shortening a line box.  If I understand Fremy correctly, he's considering the situation in which there's an overflowing unbreakable word immediately followed (in the source) by a 0-width float, and he's questioning whether the float should remain on the same line as the word, or whether it should drop down to align with the sunsequent line box.  In this case, the float can't stay on the same line as the word, because given that the line box is regarded as shortened and is not sufficiently wide to contain the word, the line box would need to be moved down below the float - and yet this contradicts the positioning rules since the float is later in the source than the word.  So the only option is to move the float down to the subsequent line.

Back to regarding the float as _not_ shortening the line box, we have to decide whether the float can stay on the same line as the word or move below.  The spec seems to have something to say here:

9.5 Floats, paragraph 7:

  # [...]. In other words, if inline-level boxes are placed on the line before a left float is encountered that fits in the remaining line box space, the left float is placed on that line, aligned with the top of the line box, and then the inline-level boxes already on the line are moved accordingly to the right of the float (the right being the other side of the left float) and vice versa for rtl and right floats.

Now, the text in this section is pretty raw and I don't think we can treat it as gospel, but my interpretation is that first we reached the word and it's found to overflow, and then we reach our 0-width left float, and we have to decide whether it "fits in the remaining line box space".  I would argue that it doesn't fit, because the word is already overflowing.  And so we don't apply the instruction to place the float on the same line.  That is, we place the float on the subsequent line.  Of course, it's also reasonable to argue that it _must_ fit because it's zero-width.  But I prefer the first interpretation, and it gets my vote.  (Plus I like the fact that it corresponds to the result that you're obliged to get if you want to regard the float as shortening the line box, thus making it a moot point whether you regard the line box as getting shortened or not.)

**Comment by tabatkins** [comment]
:( This is the problem with defining layout in terms of constraints, instead of as an algorithm - degenerate cases are often not well-defined.

While my reading of CSS2.2 is that this case is underdefined, I agree with Anton in how it *should* behave - if the linebox of preceding content is already forced to be *larger* than the available space, we should consider it as there being no space left on that line, so the float has to move down to the subsequent linebox.  (Note that this is [how Flexbox explicitly works](https://drafts.csswg.org/css-flexbox/#algo-line-break) - it collects things into a line until the *next* item would exceed the available space. So if an item takes up exactly the amount of space, we'll continue collecting zero-size items onto the line, but if an item takes up more than the amount of space, following zero-size items will be on the next line instead.)

**Comment by FremyCompany** [comment]
Resolved today that a float, whatever its width (even 0px), does count as shortening the line box. If a line box is already overflowing, the float will get pushed down to the next line.

**Comment by antonprowse** [comment]
On the last telecon it was indicated that dbaron and TabAtkins felt that the spec should be clarified to make it more obvious that _a 0px width float that is next to a line box does count as shortening a line box_ - which is something that is already a direct consequence of what the spec says, since _any_ float with a margin area of positive height that's on the same horizontal level as a line box is regarded as shortening the line box.  (The actual resolution [https://lists.w3.org/Archives/Public/www-style/2016Oct/0130.html] was that phrase which I have italicized, which is weird because it's not really a resolution; what's the action item?  Is it to make a clarification?  Or is it merely saying that the current spec is fine?  Personally I don't think a clarification is needed.  The spec is clear.)

Or to put it another way, the answer to Fremy's question at the end of the post with which he opened the issue is that we should file bugs on browsers here, rather than change the rules.

Still, since we're discussing these cases, I'll point out that I'm not sure that we should be so happy with the current rules in the spec.

Consider the opposite case where a 0-width float precedes an overflowing inline (which I mentioned in my previous post), such as in the the last two examples in https://lists.w3.org/Archives/Public/www-style/2010Aug/0182.html as well as in the following example:

&#x3C;!DOCTYPE html&#x3E;
&#x3C;div style=&#x22;width:100px; height:100px; background:gray&#x22;&#x3E;
    &#x3C;div style=&#x22;float:left; width:0; height:50px;
background:green&#x22;&#x3E;&#x3C;/div&#x3E;
    &#x3C;span&#x3E;supercalifragilisticexpialidocious&#x3C;/span&#x3E;
&#x3C;/div&#x3E;

The spec says that the inline should get pushed down to below the float (because the line box is shortened since the float is next the line box; note that the spec basically says that "next to" means you can draw a horizontal line to connect the float's margin area box to the line box, irrespective of whether they even touch - although I do suspect that that definition was arrived at in order to solve the 0-height float problem of Issue 185 [https://wiki.csswg.org/spec/css2.1#issue-185] rather than because of some deeper fundamental consideration).

But Gecko (for example) doesn't push the word down, which I think is reasonable of it; and indeed I certainly dislike treating the line box as being shortened in the case where the right margin of a preceding float is negative and doesn't even intersect the content area of the element which generates the block container box in which the line box sits - yet that's what the spec demands.

So perhaps we *should* consider changing the spec to approve Gecko's current behaviour in the latter situation.

It would make for some hard editorial work though, since we do still want to achieve the motivation for the resolution, namely that /subsequent/ 0-width floats should be pushed down to below the overflowing line box; and if we're no longer able to achieve that in the way that the spec currently does it (by saying that such a float is next to the line box and hence shortens the line box) then, since it will be horrible to try to exclude such floats from being regarded as "next to the line box", we will instead have to introduce float placeholders and say that if a placeholder is encountered immediately after an inline which overflows then it is treated as sitting at the start of the next line box instead of at the end of the one containing the overflowing inline.

Thoughts?
